added support for Rouge syntax highlighting

marianposaceanu opened the pull request:
I use [Rouge](https://github.com/jayferd/rouge) in my Rails app so instead of installing Python and the pygments gem I forked the lib and added support for Rouge.

It also makes the `controller_helpers.rb` file a bit cleaner. 

Btw. I don't see any tests/specs ?

marianposaceanu committed:
added support for Rouge syntax highlighting

dewski commented on lib/peek/rblineprof/controller_helpers.rb:
```diff
@@ -15,13 +11,9 @@ module ControllerHelpers
-      def pygmentize(file_name, code, lexer = nil)
-        if pygmentize? && lexer.present?
-          Pygments.highlight(code, :lexer => lexer_for_filename(file_name))
+      def pygmentize(file_name, code, lexer = false)
```
This could just be renamed to `highlight`

dewski commented on lib/peek/rblineprof/controller_helpers.rb:
```diff
@@ -15,13 +11,9 @@ module ControllerHelpers
-          Pygments.highlight(code, :lexer => lexer_for_filename(file_name))
+      def pygmentize(file_name, code, lexer = false)
+        if lexer
+          SyntaxHighlighter.pygmentize(code, lexer_for_filename(file_name))
```
Same method renaming here.

dewski commented on lib/peek/rblineprof/syntax_highlighter.rb:
```diff
@@ -0,0 +1,34 @@
+
+      private
+
+      def self.highlighter
```
This method is super busy, what I'd rather see is something like:

``` ruby
module Peek
  module Rblineprof
    class SyntaxHighlighter
      class_attribute :highlighter
    end
  end
end

begin
  require 'pygments.rb'
  Peek::Rblineprof::SyntaxHighlighter.highlighter = Peek::Rblineprof::SyntaxHighlighter::PygmentsHighlighter
rescue LoadError
  # Doesn't have pygments.rb installed
end

if Peek::Rblineprof::SyntaxHighlighter.highlighter.nil?
  begin
    require 'rouge'
    Peek::Rblineprof::SyntaxHighlighter.highlighter = Peek::Rblineprof::SyntaxHighlighter::RougeHighlighter
  rescue LoadError
    # Doesn't have rouge installed
  end
end
```

dewski commented on lib/peek/rblineprof/highlighters/rouge_highlighter.rb:
```diff
@@ -0,0 +1,17 @@
+        }
+
+        formatter = Rouge::Formatters::HTML.new(:line_numbers => true)
+        lexer     = lexers[lexer.to_sym].new
```
What happens if there is no lexer? Should probably return the code if the lexer key doesn't exist.

marianposaceanu committed:
Better method naming and more stable logic :
- changed pygmentize to highlight
- returs code in RougeHighlither if no lexer is found

marianposaceanu committed:
Extracted require blocks from SyntaxHighlighter class
- similar to the code in the comment
- enclosed logic into same namespace (no long lines of code)
- small typo in RougeHighlighter

lucasmazza commented:
@dewski any plans on still merging this? The patch looks interesting, specially as how Rouge evolved since 2013 :eyes:

rymai commented:
@dewski Friendly ping, any chance to get this merged at some point? :)

rymai commented on lib/peek/rblineprof/controller_helpers.rb:
```diff
@@ -120,7 +112,7 @@ def inject_rblineprof
             end
             output << "<pre class='duration'>#{times.join("\n")}</pre>"
-            output << "<div class='code'>#{pygmentize(file_name, code.join, 'ruby')}</div>"
+            output << "<div class='code'>#{highlight(file_name, code.join, true)}</div>"
```
I think it would be great to use a `<pre>` instead of a `<div>` here, or at least allow to customize that.

rymai replied:
I submitted https://github.com/peek/peek-rblineprof/pull/16 for this proposal.

dewski commented:
Hey! Sorry for the delayed response, any chance we can get the conflicts resolved? Then I'll merge.

dewski commented on lib/peek/rblineprof/syntax_highlighter.rb:
```diff
@@ -0,0 +1,38 @@
+require 'peek/rblineprof/highlighters/pygments_highlighter.rb'
+require 'peek/rblineprof/highlighters/rouge_highlighter.rb'
```
Can we remove the `.rb` suffix?

marianposaceanu committed:
Removed unnecessary extension from require

marianposaceanu committed:
resolved PR #5 merge conflicts

rymai commented on lib/peek/rblineprof/controller_helpers.rb:
```diff
@@ -17,13 +13,9 @@ module ControllerHelpers
-          Pygments.highlight(code, :lexer => lexer_for_filename(file_name))
+      def highlight(file_name, code, lexer = false)
+        if lexer
+          SyntaxHighlighter.highlight(code, lexer_for_filename(file_name))
```
I would suggest passing `SyntaxHighlighter.highlight(file_name, code)` and let `SyntaxHighlighter` handle the `lexer_for_filename(file_name)` part since:

- it's not the responsibility of the controller helper
- it will allow to monkey-patch `SyntaxHighlighter` more easily

rymai commented on lib/peek/rblineprof/syntax_highlighter.rb:
```diff
@@ -0,0 +1,38 @@
+module Peek
+  module Rblineprof
+    class SyntaxHighlighter
+      class_attribute :highlighter
```
Could we make `SyntaxHighlighter.highlighter` lazy so that we can cleanly set it to a custom highlighter in an initializer? We have this need for GitLab where we're using a custom `Gitlab::Highlight` class.

rymai commented on lib/peek/rblineprof/syntax_highlighter.rb:
```diff
@@ -0,0 +1,38 @@
+    class SyntaxHighlighter
+      class_attribute :highlighter
+
+      def self.highlight(code, lexer)
```
Here we'd receive `file_name, code` and pass them to `highlighter.process`.

We could make both highlighters inherit from an abstract class in which we could put the `lexer_for_filename(file_name)` method to stay DRY?

rymai reviewed (commented):
Thanks for your work, I'd love to be able to define a custom highlighter without any monkey-patching so I posted a few suggestions.
